src/translator/sentence_ranges.h

```diff
@@ -3,50 +3,120 @@
+/// ByteRange stores indices for begin and end (inclusive) in a string. Can be
+/// used to represent a sentence, word.
+struct ByteRange {
+  size_t begin_byte_offset;
+  size_t end_byte_offset;
+  const size_t size() const { return end_byte_offset - begin_byte_offset + 1; }
+};
+
+/// An Annotation is a collection of ByteRanges used to denote ancillary
+/// information of sentences and words on a blob of string. Annotation is meant
+/// for consumption on platforms where string_view creates problems (eg: exports
+/// through WASM). See AnnotatedBlob for cases where this is a non-issue.
+class Annotation {
 public:
-  typedef std::vector<string_view>::iterator WordIterator;
+  /// Annotation is constructed empty. See addSentence to populate it with
+  /// annotations.
+  Annotation() {}
 
-  void addSentence(std::vector<string_view> &wordRanges);
-  void addSentence(WordIterator begin, WordIterator end);
+  /// Returns the number of sentences annotated in a blob.
+  size_t numSentences() const { return sentenceBeginIds_.size(); }
 
-  void clear() {
-    flatByteRanges_.clear();
-    sentenceBeginIds_.clear();
-  }
+  /// Returns number of words in the sentece identified by sentenceIdx.
+  size_t numWords(size_t sentenceIdx) const;
 
-  size_t numSentences() const { return sentenceBeginIds_.size(); }
+  /// Adds a sentences from vector<ByteRange> representation, internally doing
+  /// extra book-keeping for the sentence terminal markings.
+  void addSentence(std::vector<ByteRange> &sentence);
+
+  /// Returns a ByteRange representing wordIdx in sentenceIdx
+  ByteRange word(size_t sentenceIdx, size_t wordIdx) const;
 
-  // Returns a string_view into the ith sentence.
-  string_view sentence(size_t index) const;
+  /// Returns a ByteRange representing sentence corresponding to sentenceIdx.
+  ByteRange sentence(size_t sentenceIdx) const;
 
 private:
-  // A flat storage for string_views. Can be words or sentences.
-  std::vector<string_view> flatByteRanges_;
+  /// A flat storage for ByteRanges. Composed of word ByteRanges, extra
+  /// information in sentenceBeginIds_ to denote sentence boundary markers as
+  /// indices.
+  std::vector<ByteRange> flatByteRanges_;
 
-  // The container grows dynamically with addSentence. size_t marking index is
-  // used to ensure the sentence boundaries stay same while underlying storage
-  // might be changed during reallocation.
+  /// Stores indices where sentences begin
   std::vector<size_t> sentenceBeginIds_;
 
-  // Utility function to extract the string starting at firstWord and ending at
-  // lastWord as a single string-view.
-  string_view sentenceBetween(string_view firstWord,
-                              string_view lastWord) const;
+  /// Returns indices of terminal (word) ByteRanges of a sentence corresponding
+  /// to sentenceIdx
+  std::pair<ByteRange, ByteRange> sentenceTerminals(size_t sentenceIdx) const;
+  std::pair<size_t, size_t> sentenceTerminalIds(size_t sentenceIdx) const;
```

Comment:
Explain what this function does and what it returns.  Every index returned by something needs to explain what it is an index into.

Reply:
```cpp
  /// Returns ByteRanges corresponding to beginning and end words of sentence
  /// corresponding to sentenceIdx. This is useful in using the information to
  /// construct a ByteRange of a sentence taking the begin from the first and
  /// end from the second.
  std::pair<ByteRange, ByteRange> sentenceTerminals(size_t sentenceIdx) const;

  /// Returns indices of terminal (word) ByteRanges in sentenceIds_ of a
  /// sentence corresponding to sentenceIdx. The distance can be used to compute
  /// number of words in a sentence (numWords) and also to construct the
  /// terminal ByteRanges (sentenceTerminals).
  std::pair<size_t, size_t> sentenceTerminalIds(size_t sentenceIdx) const;
```